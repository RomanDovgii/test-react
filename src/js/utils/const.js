export const MAXIMUM_USERS_PER_PAGE = 50;

export const AppRoute = {
  SELECT: `/select`,
  TABLE: `/table`
};
