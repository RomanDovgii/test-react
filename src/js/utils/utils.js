export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const sortToTop = (a, b) => {
  switch (true) {
    case a > b:
      return -1;
    case a < b:
      return 1;
    default:
      return 0;
  }
};

export const sortToBot = (a, b) => {
  switch (true) {
    case a < b:
      return -1;
    case a > b:
      return 1;
    default:
      return 0;
  }
};

export const filterSimpleObject = (object, searchString) => {
  let isIncluded = false;

  Object.values(object).forEach((value) => {
    isIncluded = isIncluded || String(value).toLowerCase().includes(searchString.toLowerCase());
  });

  return isIncluded;
};

export const filterComplexObject = (object, searchString) => {
  let isIncluded = false;

  Object.values(object).forEach((value) => {
    switch (typeof value) {
      case `object`:
        isIncluded = isIncluded || filterSimpleObject(value, searchString);
        break;
      default:
        isIncluded = isIncluded || String(value).toLowerCase().includes(searchString.toLowerCase());
        break;
    }
  });

  return isIncluded;
};

export const noop = () => {};
