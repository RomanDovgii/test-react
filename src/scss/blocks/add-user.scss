.add-user {
  border-collapse: collapse;
  width: 100%;
  max-width: 1340px;
}

.add-user__heading {
  border: 2px solid #333333;

  padding: 10px 10px 10px 10px;
}

.add-user__data {
  border: 2px solid #333333;
}

.add-user__input {
  box-sizing: border-box;
  border: none;

  width: 100%;
  height: 100%;

  padding: 10px 10px 10px 10px;

  &--error {
    background-color: #aa4c4c86;
    color: #eeeeee;
  }
}
