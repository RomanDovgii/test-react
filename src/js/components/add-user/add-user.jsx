import React, {createRef, Fragment} from "react";
import AddUserHeading from "../add-user-heading/add-user-heading";
import {addUserType} from "../../types/types";
import {connect} from "react-redux";
import {UNKNOWN_INFORMATION, DESCRIPTION_FOR_NEW_USER} from "../../utils/const";
import {addUser, updateCurrentUsers, updatePagesCount} from "../../store/action/action";

const AddUser = ({isOpen, isLoading, onAddClick, onCloseClick, onAddToTableClick}) => {
  const idRef = createRef();
  const firstNameRef = createRef();
  const lastNameRef = createRef();
  const emailRef = createRef();
  const phoneRef = createRef();

  return (
    <Fragment>
      {!isOpen &&
        <button className="controls__add-button" onClick={onAddClick}>Добавить</button>
      }
      {isOpen &&
        <button className="controls__add-button" onClick={onCloseClick}>Закрыть</button>
      }
      {isOpen &&
      <form className="controls__add-user">
        <table className="controls__add-user add-user">
          <thead>
            <AddUserHeading/>
          </thead>
          <tbody>
            <tr className="add-user__row">
              <td className="add-user__data">
                <input type="text" name="id" className="add-user__input" placeholder="Введите id" ref={idRef}/>
              </td>
              <td className="add-user__data">
                <input type="text" name="firstName" className="add-user__input" placeholder="Введите Имя" ref={firstNameRef}/>
              </td>
              <td className="add-user__data">
                <input type="text" name="lastName" className="add-user__input" placeholder="Введите Фамилию" ref={lastNameRef}/>
              </td>
              <td className="add-user__data">
                <input type="email" name="email" className="add-user__input" placeholder="Введите email" ref={emailRef}/>
              </td>
              <td className="add-user__data">
                <input type="text" name="phone" className="add-user__input" placeholder="Введите телефон" ref={phoneRef}/>
              </td>
            </tr>
          </tbody>
        </table>

        <button
          className="controls__add-button controls__add-button--to-table"
          type="submit"
          onClick={(evt) => {
            evt.preventDefault();
            const idValue = parseInt(idRef.current.value, 0);
            const firstNameValue = firstNameRef.current.value;
            const lastNameValue = lastNameRef.current.value;
            const emailValue = emailRef.current.value;
            const phoneValue = phoneRef.current.value;


            if (!isLoading && idValue && firstNameValue && lastNameValue && emailValue && phoneValue) {
              const user = {
                id: idValue,
                firstName: firstNameValue,
                lastName: lastNameValue,
                email: emailValue,
                phone: phoneValue,
                address: {
                  streetAddress: UNKNOWN_INFORMATION,
                  city: UNKNOWN_INFORMATION,
                  state: UNKNOWN_INFORMATION,
                  zip: UNKNOWN_INFORMATION
                },
                description: DESCRIPTION_FOR_NEW_USER
              };

              onCloseClick();
              onAddToTableClick(user);
            }
          }
          }
        >
          Добавить в таблицу
        </button>
      </form>
      }
    </Fragment>
  );
};

AddUser.propTypes = addUserType;

const mapStateToProps = (state) => ({
  isLoading: state.isLoading
});

const mapDispatchToPorps = (dispatch) => ({
  onAddToTableClick(user) {
    dispatch(addUser(user));
    dispatch(updateCurrentUsers());
    dispatch(updatePagesCount());
  }
});

export {AddUser};
export default connect(mapStateToProps, mapDispatchToPorps)(AddUser);
