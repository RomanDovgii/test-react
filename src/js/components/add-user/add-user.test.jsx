import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import {AddUser} from "./add-user";
import {noop} from "../../utils/utils";

it(
    `Should render correct closed AddUser with locked button`,
    () => {
      const renderer = new ShallowRenderer();

      const tree = renderer
      .render(
          <AddUser
            onAddClick={noop}
            onCloseClick={noop}
            onInputChange={noop}
            onAddToTableClick={noop}
            isOpen={false}
            areFilled={false}
            isLoading={true}
          />
      );

      expect(tree).toMatchSnapshot();
    }
);

it(
    `Should render correct open AddUser with locked button`,
    () => {
      const renderer = new ShallowRenderer();

      const tree = renderer
      .render(
          <AddUser
            onAddToTableClick={noop}
            isOpen={true}
            areFilled={false}
            isLoading={true}
            onAddClick={noop}
            onCloseClick={noop}
            onInputChange={noop}
          />
      );

      expect(tree).toMatchSnapshot();
    }
);

it(
    `Should render correct open AddUser with unlocked button`,
    () => {
      const renderer = new ShallowRenderer();

      const tree = renderer
      .render(
          <AddUser
            onAddToTableClick={noop}
            isOpen={true}
            areFilled={true}
            isLoading={true}
            onAddClick={noop}
            onCloseClick={noop}
            onInputChange={noop}
          />
      );

      expect(tree).toMatchSnapshot();
    }
);
