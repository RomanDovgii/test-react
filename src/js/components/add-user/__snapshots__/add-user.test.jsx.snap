// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correct closed AddUser with locked button 1`] = `
<React.Fragment>
  <button
    className="controls__add-button"
    onClick={[Function]}
  >
    Добавить
  </button>
</React.Fragment>
`;

exports[`Should render correct open AddUser with locked button 1`] = `
<React.Fragment>
  <button
    className="controls__add-button"
    onClick={[Function]}
  >
    Закрыть
  </button>
  <form
    className="controls__add-user"
  >
    <table
      className="controls__add-user add-user"
    >
      <thead>
        <AddUserHeading />
      </thead>
      <tbody>
        <tr
          className="add-user__row"
        >
          <td
            className="add-user__data"
          >
            <input
              className="add-user__input"
              name="id"
              onChange={[Function]}
              placeholder="Введите id"
              type="text"
            />
          </td>
          <td
            className="add-user__data"
          >
            <input
              className="add-user__input"
              name="firstName"
              onChange={[Function]}
              placeholder="Введите Имя"
              type="text"
            />
          </td>
          <td
            className="add-user__data"
          >
            <input
              className="add-user__input"
              name="lastName"
              onChange={[Function]}
              placeholder="Введите Фамилию"
              type="text"
            />
          </td>
          <td
            className="add-user__data"
          >
            <input
              className="add-user__input"
              name="email"
              onChange={[Function]}
              placeholder="Введите email"
              type="email"
            />
          </td>
          <td
            className="add-user__data"
          >
            <input
              className="add-user__input"
              name="phone"
              onChange={[Function]}
              placeholder="Введите телефон"
              ype="text"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <button
      className="controls__add-button controls__add-button--to-table"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Добавить в таблицу
    </button>
  </form>
</React.Fragment>
`;

exports[`Should render correct open AddUser with unlocked button 1`] = `
<React.Fragment>
  <button
    className="controls__add-button"
    onClick={[Function]}
  >
    Закрыть
  </button>
  <form
    className="controls__add-user"
  >
    <table
      className="controls__add-user add-user"
    >
      <thead>
        <AddUserHeading />
      </thead>
      <tbody>
        <tr
          className="add-user__row"
        >
          <td
            className="add-user__data"
          >
            <input
              className="add-user__input"
              name="id"
              onChange={[Function]}
              placeholder="Введите id"
              type="text"
            />
          </td>
          <td
            className="add-user__data"
          >
            <input
              className="add-user__input"
              name="firstName"
              onChange={[Function]}
              placeholder="Введите Имя"
              type="text"
            />
          </td>
          <td
            className="add-user__data"
          >
            <input
              className="add-user__input"
              name="lastName"
              onChange={[Function]}
              placeholder="Введите Фамилию"
              type="text"
            />
          </td>
          <td
            className="add-user__data"
          >
            <input
              className="add-user__input"
              name="email"
              onChange={[Function]}
              placeholder="Введите email"
              type="email"
            />
          </td>
          <td
            className="add-user__data"
          >
            <input
              className="add-user__input"
              name="phone"
              onChange={[Function]}
              placeholder="Введите телефон"
              ype="text"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <button
      className="controls__add-button controls__add-button--to-table"
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      Добавить в таблицу
    </button>
  </form>
</React.Fragment>
`;
